lock-api/src/main/java/com/palantir/lock/client/CoalescingTransactionStarter.java

```diff
@@ -0,0 +1,204 @@
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+import com.google.common.collect.Streams;
+import com.palantir.atlasdb.autobatch.BatchElement;
+import com.palantir.atlasdb.autobatch.DisruptorAutobatcher;
+import com.palantir.common.base.Throwables;
+import com.palantir.lock.v2.LockImmutableTimestampResponse;
+import com.palantir.lock.v2.LockToken;
+import com.palantir.lock.v2.PartitionedTimestamps;
+import com.palantir.lock.v2.StartIdentifiedAtlasDbTransactionResponse;
+import com.palantir.lock.v2.StartTransactionResponseV4;
+import com.palantir.lock.v2.TimestampAndPartition;
+
+/**
+ * A service responsible for coalescing multiple start transaction calls into a single start transactions call. This
+ * service also handles creating {@link LockTokenShare}'s to enable multiple transactions sharing a single immutable
+ * timestamp.
+ *
+ * Callers of this class should use {@link #unlock(Set)} and {@link #refreshLockLeases(Set)} for returned lock tokens,
+ * rather than directly calling delegate lock service.
+ */
+final class CoalescingTransactionStarter implements AutoCloseable {
+    private final DisruptorAutobatcher<Void, StartIdentifiedAtlasDbTransactionResponse> autobatcher;
+    private final LockLeaseService lockLeaseService;
+
+    private CoalescingTransactionStarter(
+            DisruptorAutobatcher<Void, StartIdentifiedAtlasDbTransactionResponse> autobatcher,
+            LockLeaseService lockLeaseService) {
+        this.autobatcher = autobatcher;
+        this.lockLeaseService = lockLeaseService;
+    }
+
+    static CoalescingTransactionStarter create(LockLeaseService lockLeaseService) {
+        return new CoalescingTransactionStarter(DisruptorAutobatcher.create(
+                consumer(lockLeaseService)),
+                lockLeaseService);
+    }
+
+    @VisibleForTesting
+    static Consumer<List<BatchElement<Void, StartIdentifiedAtlasDbTransactionResponse>>> consumer(
+            LockLeaseService lockLeaseService) {
+        return batch -> {
+            int numTransactions = batch.size();
+
+            List<StartIdentifiedAtlasDbTransactionResponse> startTransactionResponses =
+                    getStartTransactionResponses(lockLeaseService, numTransactions);
+
+            for (int i = 0; i < numTransactions; i++) {
+                batch.get(i).result().set(startTransactionResponses.get(i));
+            }
+        };
+    }
+
+    private static List<StartIdentifiedAtlasDbTransactionResponse> getStartTransactionResponses(
+            LockLeaseService lockLeaseService, int numberOfTransactions) {
+        List<StartIdentifiedAtlasDbTransactionResponse> result = new ArrayList<>();
+        while (result.size() < numberOfTransactions) {
```

Comment:
Seems slightly subtle that a single batch request may return fewer transactions than requested. Is this the right layer to handle that vs. `LockLeaseService`? Similar to `TimestampService#getFreshTimestamps`, let's document appropriately (on `TimelockService`?) that this is the case.

Reply:
I think `TimelockRpcClient` is the right place to document this, as `TimelockService` actually do not have a `startTransactions` method - and shouldn't have as it requires specific logic to handle sharing immutable timestamp lock token.
I am a bit concerned about moving the handling less-than-expected case to a lower layer, as this is the only place that requires this guarantee. We may end up using some other delegate at some point (rather than using lockLeaseService), in that case, we will need to make sure that new thing handles it correctly.